templates/usr/local/sbin/simp_generate_types.epp

```diff
@@ -3,132 +3,259 @@
+      fh.flock(File::LOCK_EX)
 
-  fh.puts(msg) if fh
+      # Never try to kill PID < 2!
+      if pid > 1
+        begin
+          Process.kill(9, pid)
 
-  exit 1
+          LOG.warn("Force unlocked '#{pid}'")
+        rescue => e
+          LOG.warn("Attempt to kill process '#{pid}' during unlock of '#{lockfile}' failed: #{e}")
+        end
+
+        File.delete(fh)
+      end
+    elsif target_pid == pid
+      fh.flock(File::LOCK_NB|File::LOCK_EX)
+      File.delete(fh)
+    end
+  end
 end
 
-output = []
+def wait_and_lock(lockfile, timeout=30)
+  if File.exist?(lockfile)
+    begin
+      Timeout::timeout(timeout) do
+        until lock(lockfile) do
+          sleep(2)
+        end
+      end
+    rescue Timeout::TimeoutError
+      unlock(lockfile)
+    end
+  end
+
+  at_exit { unlock(lockfile, Process.pid) }
 
-current_umask = File.umask
-File.umask(0022)
+  lock(lockfile)
+end
+
+def log_err(msg, options)
+  $stderr.puts Array(msg).join("\n") unless options.quiet
 
-delay = options.delay/options.target_environments.count.to_f
+  log(msg, options)
+end
 
-options.target_environments.each do |env|
-  sleep(delay)
+def log(msg, options)
+  msg = Array(msg)
 
-  # Prevent running on the same environment simultaneously
-  next unless %x{ps --no-headers -o args -wwC #{puppet_cmd}}.
-                lines.
-                grep(/generate types/).
-                grep(/--environment\s+#{env[:env_name]}(\s+|$)/).
-                empty?
+  @logfile ||= File.open(options.logfile, 'a+') if options.logfile
 
-  begin
-    output += %x{#{puppet_cmd} generate types --environment #{env[:env_name]} 2>&1}.strip.lines
-  ensure
-    resource_types_dir = File.join(env[:path], '.resource_types')
-
-    if File.exist?(resource_types_dir)
-      FileUtils.chown_R(nil, puppet_group, resource_types_dir)
-      FileUtils.chmod_R('g+r,o-rwx', resource_types_dir)
-      FileUtils.chmod('g+x', resource_types_dir)
+  @logfile.puts(msg.join("\n")) if @logfile
+
+  msg.each do |line|
+    LOG.error(line) if options.syslog
+  end
+end
+
+def process_paths(to_process, options)
+  to_process.map!(&:strip)
+
+  target_environments = []
+
+  environment_paths = Dir.glob(
+    %x{#{PUPPET_CMD} config print --log_level=err --section=master environmentpath}.strip.split(':').map{|x| x + '/*'}
+  )
+
+  if to_process.any?{|path| path.casecmp('all')==0}
+    target_environments = environment_paths.compact
+  else
+    to_process.each do |path|
+      if environment_paths.include?(path) || environment_paths.any?{|env_path| path.start_with?("#{env_path}/")}
+        target_environments << path
+      end
     end
+
+    to_process = to_process.sort.uniq
+  end
+
+  if target_environments.empty?
+    msg = 'Error: Could not find any target environments'
+    log_err(msg, options)
+    exit 1
+  end
+
+  # Convert each discovered environment into a hash noting the path and environment name
+  target_environments.map!{|e| { :path => e, :env_name => File.basename(e) }}
```

Comment:
If an old incron rule that points to a module's types directory,  File.basename(e) will not be the enviroment name.

Reply:
@lnemsick-simp Old incron rules will be erased with this module update. I can re-add that code if you like though.

Reply:
The new default rules don't have the types directory in them, but there is nothing to prevent a user from configuring for such a directory.